twine/repository.py

```diff
@@ -183,21 +183,31 @@ def upload(
     ) -> requests.Response:
         number_of_redirects = 0
         while number_of_redirects < max_redirects:
-            resp = self._upload(package)
-
-            if resp.status_code == requests.codes.OK:
-                return resp
-            if 500 <= resp.status_code < 600:
+            try:
+                resp = self._upload(package)
+            except (
+                requests.exceptions.ConnectionError,
+                requests.exceptions.ConnectTimeout,
+            ) as exc:
                 number_of_redirects += 1
                 logger.warning(
-                    f'Received "{resp.status_code}: {resp.reason}"'
+                    f"{exc.__class__.__name__} raised."
                     "\nPackage upload appears to have failed."
                     f" Retry {number_of_redirects} of {max_redirects}."
                 )
+                continue
             else:
-                return resp
-
-        return resp
+                if resp.status_code == requests.codes.OK:
```

Comment:
slighlty redundant, but can leave for clarity. Should add a continue to the retry if clause.

Reply:
Additionally, as the _upload function calls on POST without redirects, we aren't really handling anywhere (they are simply returned)